pam/internal/adapter/pwquality_c.go

```diff
@@ -0,0 +1,46 @@
+package adapter
+
+// #cgo pkg-config: pwquality
+// #include <stdio.h>
+// #include <stdlib.h>
+// #include <pwquality.h>
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"unsafe"
+)
+
+// checkChallengeQuality checks the quality of the new password using the pwquality library.
+func checkChallengeQuality(old, new string) error {
+	pwq := C.pwquality_default_settings()
+	if pwq == nil {
+		return errors.New("could not allocate pw quality default settings")
+	}
+	defer C.pwquality_free_settings(pwq)
+
+	var auxErr *C.char
+	auxErrPointer := unsafe.Pointer(auxErr)
+	defer C.free(auxErrPointer)
```

Comment:
I think this is wrong, because we'll end up in a [double free](https://github.com/libpwquality/libpwquality/blob/80deb29da767961676fedd87dbf264acc438a720/src/error.c#L118-L143).

In fact the pointer has to be completely managed by libpwquality, we only have to keep it around and pass it.

In case that's set, then it's up to `pwquality_strerror` to free it when it's set in a way that it needs too. It doesn't look a great design of the library to me, but that's what [it does](https://github.com/libpwquality/libpwquality/blob/80deb29da767961676fedd87dbf264acc438a720/src/pwscore.c#L65-L69) too.

Or if we want to be simpler just return the `errMsg` and that's it.

Reply:
Well, it's definitely not a good design, but we are not crashing so I think the compiler is doing something with this `free`. I wouldn't remove it, as it can be useful in cases where the dependency might actually leak memory.

Reply:
No, really we should not touch that pointer.

The reason we don't crash yet is that it's always `nil` when we hit the defer function, but if you end up in one of those cases we will.

Also [doc](https://github.com/libpwquality/libpwquality/blob/master/doc/man/pwquality.3.pod) is clear:

> Function pwquality_strerror() translates the errcode and auxerror auxiliary data into a localized text message. If buf is NULL the function uses an internal static buffer which makes the function non-reentrant in that case. The returned pointer is not guaranteed to point to the buf. The function deallocates eventual auxerror data passed into it, thus it must not be called twice with the same auxerror data.